### ISSUE TITLE
Exit code 0xffffffff (-1) for (butler) / Segmentation fault on command-line run

### ISSUE THREAD

**Issue by hsandt** [itchio/butler#189]
**Butler v15.12.0**

When running butler from itch.io, I get:

> Exit code 0xffffffff (-1) for (butler)

and from the command-line:

> Segmentation fault

I cannot copy the full log (only itch.io shows more info than "Segmentation Fault", but the app prevents text highlighting) but the report should have been to the developers since I checked the box at the bottom (on the screenshot, in German, it says I'm sending the report for bug resolution). I'm still opening this issue to follow the resolution.

![image](https://user-images.githubusercontent.com/1067772/57724348-832b7480-7682-11e9-8a7b-7369c9f0e753.png)

**Comment by hsandt** [comment]
I noticed 2 things:
1. itch.io app seems to offer all versions of butler. But in practice, when you click on any download button on Ubuntu, you only have the choice between butler linux 64-bit (bleeding-edge) and linux 32-bit (bleeding-edge). So I can't get stable even if I want to from inside the app.

![image](https://user-images.githubusercontent.com/1067772/57725276-a820e700-7684-11e9-89ae-a6966ab98098.png)

The linux 64-bit (bleeding-edge) was crashing for me, it was the one that showed the issue above. I downloaded linux 32-bit (bleeding-edge), still inside the itch.io app, and it worked fine. Then I _redownloaded_ linux 64-bit (bleeding-edge) and this time it worked fine.

2. I went to the website to download directly the files, and all versions (32/64, stable and bleeding-edge) work.

So my conclusion is that:
- itch.io app installed the bleeding-edge of butler (linux 64-bit) without me knowing.
- it worked until a certain point where it must have upgraded to a version that caused Segmentation Fault indeed
- since then the issue was fixed (on purpose or due to some side effect like a cleanup), so any new bleeding-edge download would get the right version

**Version sum-up**

64-bit bleeding-edge
head, built on Apr 30 2019 @ 17:06:59, ref 55793fadd2a5b161a46c12b5faac8aee1b5ac6b6

64-bit stable
v15.12.0, built on Apr 27 2019 @ 01:40:42, ref 4baabbcdc78269e996c3eaef3ec8da4c72c29aac

I installed butler for the NaNoRenO around April 1st, so the crash and fix must have happened in April, although I don't see anything matching by looking at the commit titles.

**Request update**

Since the issue was fixed, I will transform my request (maybe I should open a new issue...):
1. Allow to download stable versions of butler from Linux itch.io app
2. Make the upgrade process from itch.io more transparent so we know which version we have and how to upgrade without downloading a new file (every time I re-download butler it will create a folder "butler 2", "butler 3", etc.).

Request 1 seems to be a missing config on the butler entry on itch.io.

Request 2 is on itch.io so I may repost it on https://github.com/itchio/itch/issues instead, unless there is already a similar issue. I search at "upgrade app" but don't see anything matching; tell me if you know an issue about the question of downloading copy vs upgrading + how to show currently installed versions.

I'll wait for developers' input so I know if I should rather close this issue and respawn it as Request 1, or just change its title.

**Comment by fasterthanlime** [comment]
Well, I'm confused about a few things.

I'm writing this from Ubuntu 19.04 64-bit, here's what I see if I try to install butler from the app:

![image](https://user-images.githubusercontent.com/7998310/57728209-a3affa80-7693-11e9-908e-114b925675d7.png)

You can also see that both the stable and the bleeding-edge version are tagged properly here: https://fasterthanlime.itch.io/butler

Installing stable works great:

![image](https://user-images.githubusercontent.com/7998310/57728284-da861080-7693-11e9-827a-788a89b0df1d.png)

(So does launching it from the itch app, but that's not very useful, since it's a command-line tool).

Installing bleeding-edge works great too:

![image](https://user-images.githubusercontent.com/7998310/57728344-02757400-7694-11e9-8450-29740caba127.png)

You'd end up with `butler`, `butler 2`, `butler 3` directories is if you installed several channels in parallel, like `butler 64-bit stable` and `butler 64-bit bleeding-edge`, that's just how the itch app names install folders (it has nothing to do with versions). I suppose another way is if there was another existing folder named `butler` that wasn't managed by the app in first place.

Upgrades to either stable or bleeding-edge are managed automatically by the app (it needs to be open though), you can find out which version you have in manage:

![image](https://user-images.githubusercontent.com/7998310/57728405-233dc980-7694-11e9-92b9-a7970b2f2cec.png)

![image](https://user-images.githubusercontent.com/7998310/57728499-6435de00-7694-11e9-82d1-6f5790113aed.png)

![image](https://user-images.githubusercontent.com/7998310/57728536-71eb6380-7694-11e9-9847-c35aea62cba7.png)

---

So, I'm not sure how you ended up with a butler binary that crashes with 0xFFFFFFFF :( All butler binaries pushed to itch.io go through unit and integration tests (on the final, release binary), so it's unlikely a bug that slipped into a few releases, and it's not fundamentally incompatible with your system either (since you can run recent builds), so my money is on a corrupted executable (maybe butler's own self-upgrade mechanism malfunctioning? although I've never had any reports of that).

**Comment by fasterthanlime** [comment]
Just thought of something re this screenshot:

![](https://user-images.githubusercontent.com/1067772/57725276-a820e700-7684-11e9-89ae-a6966ab98098.png)

I can think of one way to obtain that:

  * Already having `butler 32-bit stable` *and* `butler 64-bit stable` installed
  * Ignoring the launch bar at the bottom:

![image](https://user-images.githubusercontent.com/7998310/57728826-17063c00-7695-11e9-90b5-e6b106a4bfc0.png)

  * and instead clicking on the embedded browser on one of those download buttons:

![image](https://user-images.githubusercontent.com/7998310/57728866-284f4880-7695-11e9-9e11-5c6fd69a3d74.png)

For example, right now, I have 64-bit bleeding-edge installed (from the previous comment), so it's excluded from that list (because it's already installed):

![image](https://user-images.githubusercontent.com/7998310/57728939-43ba5380-7695-11e9-9f68-e4d10b793c16.png)
